internal/tasks/fleet_selector.go

```diff
@@ -0,0 +1,317 @@
+package tasks
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/store"
+	"github.com/flightctl/flightctl/internal/store/model"
+	"github.com/flightctl/flightctl/internal/util"
+	"github.com/flightctl/flightctl/pkg/log"
+	"github.com/flightctl/flightctl/pkg/reqid"
+	"github.com/go-chi/chi/v5/middleware"
+	"github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+/*
+Wait to be notified via channel about fleet template updates, exit upon ctx.Done()
+The logic assumes that we deal with one update at a time. Later, we can increase scale
+by dealing with one update per org at a time.
+*/
+func FleetSelectorMatching(taskManager TaskManager) {
+	for {
+		select {
+		case resourceRef := <-taskManager.channels[ChannelFleetSelectorMatching]:
+			requestID := reqid.NextRequestID()
+			ctx := context.WithValue(context.Background(), middleware.RequestIDKey, requestID)
+			log := log.WithReqIDFromCtx(ctx, taskManager.log)
+			logic := FleetSelectorMatchingLogic{
+				taskManager: taskManager,
+				ctx:         ctx,
+				log:         log,
+				fleetStore:  taskManager.store.Fleet(),
+				devStore:    taskManager.store.Device(),
+				resourceRef: resourceRef,
+			}
+
+			if resourceRef.Kind == model.FleetKind {
+				err := logic.fleetSelectorUpdated()
+				if err != nil {
+					taskManager.log.Errorf("failed checking device ownership %s/%s: %v", resourceRef.OrgID, resourceRef.Name, err)
+				}
+			} else if resourceRef.Kind == model.DeviceKind {
+				err := logic.deviceLabelsUpdated()
+				if err != nil {
+					taskManager.log.Errorf("failed checking device ownership %s/%s: %v", resourceRef.OrgID, resourceRef.Name, err)
+				}
+			} else {
+				taskManager.log.Errorf("FleetSelectorMatching called with unexpected resource kind %s", resourceRef.Kind)
+			}
+		case <-taskManager.ctx.Done():
+			taskManager.log.Info("Received ctx.Done(), stopping")
+			return
+		}
+	}
+}
+
+type FleetSelectorMatchingLogic struct {
+	taskManager TaskManager
+	ctx         context.Context
+	log         logrus.FieldLogger
+	fleetStore  store.Fleet
+	devStore    store.Device
+	resourceRef ResourceReference
+}
+
+func (f FleetSelectorMatchingLogic) fleetSelectorUpdated() error {
```

Comment:
We would need to architect all this in a way that we could additionally perform dry-run checks, i.e. from a pull request/merge request that we have been notified of via webhook.

Reply:
Not sure I follow, let's discuss on Monday.